lib/SILOptimizer/Mandatory/ConstantPropagation.cpp

```diff
@@ -985,10 +1032,297 @@ constantFoldStringConcatenation(ApplyInst *AI,
+  // literal or an expression.
+  auto *IntLitExp = getIntegerLiteralExprFromConstructorCall(AI);
+
+  // Conversion of negative values to unsigned types is not allowed.
+  if (!DstTySigned && V.isNegative()) {
+    SmallString<10> SrcAsString;
+    V.toString(SrcAsString, /*radix*/10, true);
+
+    if (IntLitExp && IntLitExp->getLoc().isValid()) {
+      diagnose(Ctx, IntLitExp->getLoc(),
+               diag::negative_integer_literal_overflow_unsigned,
+               AI->getSubstitutions()[0].getReplacement(), SrcAsString);
+      return false;
+    }
+
+    diagnose(Ctx, AI->getLoc().getSourceLoc(),
+             diag::integer_conversion_sign_error,
+             AI->getSubstitutions()[0].getReplacement());
+    return false;
+  }
+
+  // OK, now we know the literal value.
+  // Check if it is within a range allowed by the type.
+  if ((DstTySigned && !V.isSignedIntN(DstTySize)) ||
+      (!DstTySigned && !V.isIntN(DstTySize))) {
+
+    SmallString<10> SrcAsString;
+    V.toString(SrcAsString, /*radix*/10, true);
+
+    if (IntLitExp && IntLitExp->getLoc().isValid()) {
+      diagnose(Ctx, IntLitExp->getLoc(),
+               diag::integer_literal_overflow_builtin_types, DstTySigned,
+               AI->getSubstitutions()[0].getReplacement(), SrcAsString);
+      return false;
+    }
+
+    diagnose(Ctx, AI->getLoc().getSourceLoc(),
+             diag::integer_conversion_overflow,
+             SrcTy, AI->getSubstitutions()[0].getReplacement());
+    return false;
+  }
+
+  // The constant value is in a domain allowed by the DstTy.
+
+  // This is where the initialized value will be stored. This can
+  // be not only the alloc_stack but e.g. an element inside of the array, etc.
+  auto ResultAddr = AI->getArgument(0);
+
+  // Check that this memory location is assigned just once (as an @out parameter
+  // of this initializer call).
+  bool OptimizeLoadsOfConst = isa<AllocStackInst>(ResultAddr);
+  if (OptimizeLoadsOfConst) {
+    for (auto Use : ResultAddr->getUses()) {
+      auto User = Use->getUser();
+      if (User == AI)
+        continue;
+      if (isa<StoreInst>(User) || isa<ApplyInst>(User)) {
+        OptimizeLoadsOfConst = false;
+        break;
+      }
+    }
+  }
+
+  // Replace this generic constructor call with a simple integer literal of a
+  // proper size.
+  SILBuilderWithScope B(LI);
+  auto NewIL = B.createIntegerLiteral(
+      AI->getLoc(), SILType::getBuiltinIntegerType(
+                        DstTySize, AI->getModule().getASTContext()),
+      DstTySigned ? V.sextOrTrunc(DstTySize) : V.zextOrTrunc(DstTySize));
+  auto NewStructI = B.createStruct(
+      AI->getLoc(), SILType::getPrimitiveObjectType(DstTy), {NewIL});
+
+  // Store into the ResultAddr.
+  B.setInsertionPoint(AI);
+  B.createStore(AI->getLoc(), NewStructI, ResultAddr,
+                StoreOwnershipQualifier::Unqualified);
+
+  auto RemoveCallback = [&](SILInstruction *DeadI) { WorkList.remove(DeadI); };
+
+  if (OptimizeLoadsOfConst) {
```

Comment:
Is this really needed? Can this be done by subsequent mandatory passes after the constructor call is folded?